thorlcr/activities/hashdistrib/thhashdistribslave.cpp

```diff
@@ -2924,12 +3079,27 @@ class HashDedupSlaveActivityBase : public CSlaveActivity
                 lastEog = false;
                 if (bucketHandler->addRow(row)) // true if new, i.e. non-duplicate (does not take ownership)
                 {
-                    dataLinkIncrement();
-                    return row.getClear();
+                    if (!keepBest)
+                    {
+                        dataLinkIncrement();
+                        return row.getClear();
+                    }
                 }
             }
             else
             {
+                if (keepBest && !keepBestRowsReadyToReturn)
+                {
+                    bucketHandlerIterator.setown(new CBucketHandlerIterator(bucketHandler) );
+                    if (bucketHandlerIterator->first())
+                    {
+                        keepBestRowsReadyToReturn = true;
+                        const void * row = bucketHandlerIterator->query();
+                        LinkThorRow(row);
+                        return row;
+                    }
+                }
+                keepBestRowsReadyToReturn = false;
                 // If spill event occurred, disk buckets + key buckets will have been created by this stage.
                 bucketHandler->flushBuckets();
```

Comment:
Ideally - as it's walking the non-spilt bucket rows, it would release rows as it goes.